joram.py

```diff
@@ -1,24 +1,46 @@
-#!/bin/python
 import re
 from io import BytesIO
 import requests
-from PyPDF2 import PdfFileReader
+from bs4 import BeautifulSoup
+import datetime
+from PyPDF2 import PdfReader
 from constants import GASOLINE_95, DIESEL, COLORED_DIESEL, PDF_GAS_PRICE_REGEX
 
-# TODO: Get url with current year
-#       Get all pdf's with the current date
-#       Search for gas prices
-#       Script will run only on friday every 1h to check for new documents
+# Get the current year
+current_year = str(datetime.datetime.now().year)
 
+# JORAM URL for current year's PDFs
+url = f"https://joram.madeira.gov.pt/joram/2serie/Ano%20de%20{current_year}/"
+
+response = requests.get(url)
+html = response.text
+soup = BeautifulSoup(html, "html.parser")
+links = soup.find_all("a")
+
+pdf_links = [link for link in links if link["href"].endswith(".pdf")]
+sorted_pdf_links = sorted(
+    pdf_links,
+    key=lambda link: datetime.datetime.strptime(
+        re.search(r"\d{4}-\d{2}-\d{2}", link["href"]).group(), "%Y-%m-%d"
+    ),
+)
+newest_pdf_link = sorted_pdf_links[-1]
+newest_pdf_filename = newest_pdf_link["href"].split("/")[-1]
```

Comment:
We need to scrape them all to filter only the ones with matches. Then sort them. 

Then we can get it exactly like this.

The current output is this one:
We need to do this because not all files have the info we need to extract.
```sh
IISerie-019-2023-01-26.pdf
{}
```

Reply:
I was wondering if this runs daily, and it keeps looking for the prices on the latest PDF, then we wouldn't have to scrape all the PDFs, just the latest one. So if it has price, we update it. If it doesn't, we don't update.

Reply:
The fuel prices are usually added to JORAM on Thursdays / Fridays. 

@HarryVasanth what happens if for some reason they add two PDFs at the same time? Only the last one will be scrapped?

Reply:
Hmm. That's a good point. 

But it's supposed to run every Friday, because JORAM updates a new file every Friday.

But yeah, it's a good solution after all 👀

Reply:
@carlosrsabreu: If it's on the same day, it will scrape the most recent one. So, one thing we could do is to increase the frequency of scraping to every 15 mins. So even if they make amendments to the price, we'll push the latest possible one. 

Currently, we are scraping at these two frequencies:

**Every 6 hours.**
```yml
cron: "0 */6 * * *"
 ```
 
**Every 30 mins on Fri, Sat, Sun, and Mon.**
```yml 
cron: "*/30 * * * 5,6,0,1"
```